### PR TITLE
Added GenomeReference and Contig classes

### PR DESCRIPTION
**Pull request by jigold**
This is a prototype for just build 37. My plan is to add other genome builds, expose as variable in HailContext, and add support for reading JSON from a file not in the Java resources in subsequent PRs.

### CODE REVIEW

File: src/main/resources/reference/human_g1k_v37.json [hunk 1]
```diff
@@ -0,0 +1,720 @@
+{
+    "contigs": [
+        {
+            "name": "1",
+            "length": 249250621,
+            "isX": false,
```
Comment (tpoterba) [inline src/main/resources/reference/human_g1k_v37.json]: Couldn't you just encode the X and Y chromosomes in lists instead of these booleans everywhere? Same with PAR

Comment (tpoterba) [inline src/main/resources/reference/human_g1k_v37.json]: unless this is some sort of reference standard

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    try {
+      if (resourceStream == null) {
+        throw new RuntimeException(s"Could not read genome reference file `$file'.")
+      }
+      val json = JsonMethods.parse(resourceStream)
+
+      val fields = json.asInstanceOf[JObject].obj.toMap
+
+      def getAndCastJSON[T <: JValue](fields: Map[String, JValue], fname: String)(implicit tct: ClassTag[T]): T =
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: this is duplicated from one of the read functions, so it could probably be abstracted

File: src/main/resources/reference/human_g1k_v37.json (hunk 1 above)
Comment (tpoterba) [inline src/main/resources/reference/human_g1k_v37.json]: ah, I see. You wanted it to match the case class definition? 
I still think it should be simplified. It should be as easy as possible for someone else to come along and write a reference json

**Comment by jigold** [comment]
Back to you.

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+import is.hail.expr.{JSONAnnotationImpex, TInterval}
+import is.hail.utils._
+import org.json4s._
+import org.json4s.jackson.JsonMethods
+
+import scala.reflect.ClassTag
+
+case class GenomeReference(name: String, contigs: Array[Contig], xContigs: Array[String],
+  yContigs: Array[String], mtContigs: Array[String], par: Interval[Locus]*) extends Serializable {
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: Why varargs for the par?  Shouldn't this just be an array?

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  def inX(contig: String): Boolean = xContigs.contains(contig)
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: O(N) search. genome reference should store sets for these contig lists.

**Comment by tpoterba** [comment]
We define a new file format here, though, so I'd like a second opinion. Can you find someone else to just look at the json as well,?

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    val resourceStream = Thread.currentThread().getContextClassLoader.getResourceAsStream(file)
+
+    try {
+      if (resourceStream == null) {
+        throw new RuntimeException(s"Could not read genome reference file `$file'.")
+      }
+
+      val json = JsonMethods.parse(resourceStream)
+
+      val fields = json.asInstanceOf[JObject].obj.toMap
```
Comment (cseed) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: I always thought the metadata code was a mess.  I think we can do better.  I'd use `JValue.extract` to load the data into a case class.  See "Extracting Values" here: http://json4s.org/.

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    } catch {
+      case npe: NullPointerException =>
+        throw new RuntimeException(s"Error while locating file $file", npe)
+      case e: Exception =>
+        throw new RuntimeException(s"Error loading data from $file", e)
+    } finally {
+      if (resourceStream != null) {
+        try {
+          resourceStream.close()
```
Comment (cseed) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: I don't believe close can throw exceptions.  Double-check the docs.

Comment (jigold) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: Java docs for InputStream say it throws an IOException

**Comment by jigold** [comment]
@cseed Back to you.

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    val resourceStream = Thread.currentThread().getContextClassLoader.getResourceAsStream(file)
+
+    try {
+      if (resourceStream == null) {
+        throw new RuntimeException(s"Could not read genome reference file `$file'.")
+      }
+
+      val json = JsonMethods.parse(resourceStream)
+
+      json.extract[JSONExtractGenomeReference].toGenomeReference
```
Comment (cseed) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: Much cleaner!  Very nice.  Digging into the json4s code, looks like we define custom extractors (for things like Interval) which we should use instead of these JSONExtract classes I created.  Will be cleaner again, but we can leave that for another PR.

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    } catch {
+      case npe: NullPointerException =>
```
Comment (cseed) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: I don't think this can happen anymore.

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+object GenomeReference {
+  def GRCh37 = fromResource("reference/human_g1k_v37.json")
+
+  def fromResource(file: String): GenomeReference = {
+    val resourceStream = Thread.currentThread().getContextClassLoader.getResourceAsStream(file)
+
+    try {
+      if (resourceStream == null) {
+        throw new RuntimeException(s"Could not read genome reference file `$file'.")
```
Comment (cseed) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: Why not just `assert(resourceStream = null, s"Could not read genome reference $file");`?

File: src/main/scala/is/hail/variant/GenomeReference.scala (hunk 7 above)
Comment (jigold) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: The problem with Interval was the implicit value `Ordering[T]` cannot be handled by json4s. https://github.com/json4s/json4s/issues/369, https://github.com/json4s/json4s/issues/139

**Comment by jigold** [comment]
@cseed I refactored the resource stream into utils. Back to you.

File: src/main/scala/is/hail/utils/package.scala [hunk 10]
```diff
@@ -473,4 +473,21 @@ package object utils extends Logging
     }
     expansion
   }
+
+  def loadFromResource[T](file: String)(reader: (InputStream) => T): T = {
```
Comment (cseed) [inline src/main/scala/is/hail/utils/package.scala]: Nice!

**Comment by cseed** [comment]
@tpoterba Can you take another look?

File: src/main/scala/is/hail/utils/package.scala [hunk 11]
```diff
@@ -473,4 +473,21 @@ package object utils extends Logging
...
+    val resourceStream = Thread.currentThread().getContextClassLoader.getResourceAsStream(file)
+    assert(resourceStream != null, s"Error while locating file `$file'")
+
+    try {
```
Comment (cseed) [inline src/main/scala/is/hail/utils/package.scala]: @jigold Just a poke, this is waiting on this suggestion from Tim.

Comment (jigold) [inline src/main/scala/is/hail/utils/package.scala]: I think this code is correct as is. The real error messages are still being printed -- they are prefixed with the Hail message. We need to close the resource in a "finally" block. For example, if `reader` throws an exception, we still want to close the resource.

Comment (tpoterba) [inline src/main/scala/is/hail/utils/package.scala]: why not a single try/finally then?

Comment (jigold) [inline src/main/scala/is/hail/utils/package.scala]: The resourceStream must be defined outside of the try/finally block. Otherwise, it's not accessible to the finally code block.

Comment (cseed) [inline src/main/scala/is/hail/utils/package.scala]: I think you should catch/ignore the exception on the close, but just propagate the exception from the reader.  As Tim points out, this an internal error, I don't see what wrapping the exception in a new RuntimeException buys us.

So something like:

```
try { read() }
finally { 
  try { close() } catch { e: Exception => }
}
```

Thoughts?

Comment (jigold) [inline src/main/scala/is/hail/utils/package.scala]: @tpoterba back to you.

File: src/main/scala/is/hail/utils/package.scala [hunk 12]
```diff
@@ -473,4 +473,19 @@ package object utils extends Logging
...
+      reader(resourceStream)
+    } finally {
+      try {
```
Comment (tpoterba) [inline src/main/scala/is/hail/utils/package.scala]: remove this try/catch

Comment (tpoterba) [inline src/main/scala/is/hail/utils/package.scala]: it should just be 
```
try {
    reader(...)
} finally {
 resourceStream.close()
}

File: src/main/scala/is/hail/utils/package.scala [hunk 13]
```diff
@@ -473,4 +473,14 @@ package object utils extends Logging
...
   }
+
+  def loadFromResource[T](file: String)(reader: (InputStream) => T): T = {
+    val resourceStream = Thread.currentThread().getContextClassLoader.getResourceAsStream(file)
+    assert(resourceStream != null, s"Error while locating file `$file'")
+
+    try
+      reader(resourceStream)
+    finally
+      resourceStream.close()
```
Comment (jigold) [inline src/main/scala/is/hail/utils/package.scala]: @danking @cseed Do you approve of these changes?
